**Preamble**

This is a **group** project. Please indicate the names of all group members below. 
Each group is to comprise *five* to *six* members.
To be clear, the group should submit ONE notebook.

Please acknowledge all sources. 

Names
- 1. Zhong Hao
- 2. Wen Xingdi
- 3.
- 4.
- 5.
- 6. 
- 7.

Sources (if applicable) :  
[networkx求解最小费用最大流并可视化数据](https://blog.csdn.net/engineoid/article/details/107371618)  
[Python小白的数学建模课-20.网络流优化案例](https://blog.csdn.net/youcans/article/details/118882174)  
[图与网络——最小费用最大流问题精解](https://www.cnblogs.com/haohai9309/p/18409962)  
[min_cost_flow — NetworkX 3.4.1 documentation](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.flow.min_cost_flow.html)  
Claude

# Minimum Cost Network Flow Problem

Project objectives:

i. Investigate the application of minimum cost network flow (MCNF) in solving real-world problems.

ii. Solve specific instances using both `PuLP` and `networkx` for comparison and analysis.

----

# Tasks

1. Choose a problem from this website: [Gurobi Jupyter Notebook Examples](https://www.gurobi.com/jupyter_models/?_difficulty_level=introductory). Note that the routines are written for the `Gurobi` library. <font color='red'>The syntax differs from that of `PuLP` 
and `networkx`</font>. You should be able to access the relevant datasets provided on `Gurobi`'s website.

2. Understand the problem statement and formulate it as an instance of a Minimum Cost Network Flow  (MCNF). Clearly describe the network and its parameters, including demands, costs, and capacities.
 - **Note**: You are allowed to _simplify_ the problem.  If you do, make sure to explicitly state your simplifying assumptions.

3. Write routines to solve the problem (using the dataset provided) using
 - the `PuLP` library: Reformulate the problem as a linear program.
 - the `networkx` library.
I have provided an oversimplified example for the **assignment problem** discussed in class. See `MCNF Oversimplified Project v3.ipynb` in this [Dropbox Folder](https://www.dropbox.com/scl/fo/3f26s4pdlc17gn99hgq15/AH_uc01lQ1UMV276xHTXXQc?rlkey=s7jneaz34gu950qwo6t1bm0gy&dl=0). Do note that I did not make any comparisons in the example notebook.

4. Compare the two solutions and discuss any differences. Consider factors such as running time and accuracy. Beyond implementation differences, you can also explore aspects like problem relaxation, sensitivity analysis, or certificates of optimality.

5. Present your findings in a **video** (maximum 20 minutes) 

# Submission Details

Your submission must include the following.
 - Provide the `PuLP` and `networkx` routines in this notebook. I will run the routines locally on my machine.
 - Prepare a short video (maximum 20 minutes) presenting your findings. In the video, you should:
    - (i) Describe the problem statement.
    - (ii) Formulate problem as an instance of the Minimum Cost Network Flow (MCNF).
    - (iii) Discuss the differences in the `PuLP` and `networkx` implementations, considering factors such as running time.
    - (iv) Share any interesting observations or insights related to optimisation, such as problem relaxation, sensitivity analysis, or certificates of optimality.
  - Please **share the link in this notebook**. <font color = 'red'>DO NOT SEND ME THE VIDEO FILE</font>. You may supplement your explanation with the pictures, slides, or any written notes in this notebook.
  - Please <font color='blue'>zip your notebook and the relevant datasets into a single file</font> for submission. 
  
On **24 Oct**, each group will have a _10-minute session_ with me, during which I will ask questions based on the video and the group will respond verbally.

## Solving the Resource Assignment Problem as a Minimum Cost Network Flow

----------------

### Problem Statement

<font color = 'blue'>Video Link: [YOUR LINK GOES HERE](https://quotefancy.com/quote/5580/C-S-Lewis-Integrity-is-doing-the-right-thing-even-when-no-one-is-watching)</font>

We will reformulate the Resource Assignment Problem from the Gurobi "Introduction to Mathematical Optimisation Modeling" [notebook](https://www.gurobi.com/jupyter_models/introduction-to-mathematical-optimization-modeling/) as an instance of a Minimum Cost Network Flow (MCNF) problem.

We have:  
Resources (Candidates): Carlos, Joe, and Monika.  
Jobs (Positions): Tester, Java Developer, and Architect.  
Matching Scores: Each candidate has a score representing their suitability for each job.  
Objective: Assign resources to jobs to maximise the total matching score.  

Constraints:  
Each job must be assigned to exactly one resource.  
Each resource can be assigned to at most one job.

Simplifying Assumptions:
- Capacities are set to 1 to represent that each resource/job can be used at most once.
- Costs are the negative of matching scores to convert the maximisation problem into a minimisation problem.
- Flows are integer values (0 or 1) due to capacity constraints.

We will then solve it using both the PuLP and NetworkX libraries.

----------------
### Formulate as an MCNF Problem

To model this as an MCNF problem, we construct a network with the following components:  

Nodes:  
- Source Node: Source  
- Resource Nodes: Carlos, Joe, Monika  
- Job Nodes: Tester, JavaDeveloper, Architect  
- Sink Node: Sink  

Edges:  
- Edge from Source to each resource with capacity = 1 and cost = 0.  
- Edge from each resource to each job with capacity = 1 and cost = -matching_score (since we aim to maximize matching scores).  
- Edge from each job to Sink with capacity = 1 and cost = 0.

Supplies/Demands:  
- Source Node: Supply of 3 units (equal to the number of jobs).  
- Sink Node: Demand of 3 units.  
- Intermediate Nodes (Resources and Jobs): Neither supply nor demand (flow conservation applies).

------------------

### `PuLP` Routine

In [1]:
# WRITE YOUR CODE HERE
# import libraries
import pulp

Data Preparation

In [2]:
# Resources and Jobs
resources = ['Carlos', 'Joe', 'Monika']
jobs = ['Tester', 'JavaDeveloper', 'Architect']

# matching scores
scores = {
    ('Carlos', 'Tester'): 53,
    ('Carlos', 'JavaDeveloper'): 27,
    ('Carlos', 'Architect'): 13,
    ('Joe', 'Tester'): 80,
    ('Joe', 'JavaDeveloper'): 47,
    ('Joe', 'Architect'): 67,
    ('Monika', 'Tester'): 53,
    ('Monika', 'JavaDeveloper'): 73,
    ('Monika', 'Architect'): 47
}

Create the Linear Program Problem

In [3]:
# define the problem
prob = pulp.LpProblem("Resource Assignment Problem as MCNF", pulp.LpMinimize)

c:\ProgramData\anaconda3\envs\mlaenv\lib\site-packages\pulp\pulp.py:1298: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


Define Variables

In [4]:
# flow variables between resources and jobs
flow_vars = {}
for r in resources:
    for j in jobs:
        var = pulp.LpVariable(f'flow_{r}_{j}', cat='Binary')
        flow_vars[(r, j)] = var

# flow variables from source to resources
source_vars = {}
for r in resources:
    var = pulp.LpVariable(f'flow_Source_{r}', cat='Binary')
    source_vars[r] = var

# flow variables from jobs to sink
sink_vars = {}
for j in jobs:
    var = pulp.LpVariable(f'flow_{j}_Sink', cat='Binary')
    sink_vars[j] = var

Objective Function

Since we want to maximize the total matching score, we minimize the negative total matching score.

In [5]:
# set the objective function
prob += pulp.lpSum([
    -scores[(r, j)] * flow_vars[(r, j)] for r in resources for j in jobs
])

Constraints

**Flow Conservation Constraints**  
At Resources:  
$$
\text{flow from Source to Resource} = \sum_{\text{Jobs}} \text{flow from Resource to Jobs}
$$
At Jobs:  
$$
\sum_{\text{Resources}} \text{flow from Resources to Job} = \text{flow from Job to Sink}
$$

In [6]:
for r in resources:
    prob += source_vars[r] == pulp.lpSum([flow_vars[(r, j)] for j in jobs]), f'Flow_Conservation_Resource_{r}'

for j in jobs:
    prob += pulp.lpSum([flow_vars[(r, j)] for r in resources]) == sink_vars[j], f'Flow_Conservation_Job_{j}'

**Capacity Constraints**  
Source to Resources and Jobs to Sink capacities are 1.  
Resources to Jobs capacities are 1 (already defined by binary variables).

In [7]:
for r in resources:
    prob += source_vars[r] <= 1, f'Capacity_Source_{r}'

for j in jobs:
    prob += sink_vars[j] <= 1, f'Capacity_Sink_{j}'

**Supply and Demand at Nodes**  
Source supplies 3 units.  
Sink demands 3 units.

In [8]:
prob += pulp.lpSum([source_vars[r] for r in resources]) == 3, 'Source_Supply'
prob += pulp.lpSum([sink_vars[j] for j in jobs]) == 3, 'Sink_Demand'

Solve the Problem

In [9]:
# solve the problem
prob.solve()

1

When prob.solve() returns 1, it means that the solver has found an optimal solution to the problem.

Results

In [10]:
print("Status:", pulp.LpStatus[prob.status])
print("Total Matching Score:", -pulp.value(prob.objective))

print("\nAssignments:")
for (r, j), var in flow_vars.items():
    if var.varValue > 0.5:
        print(f"{r} assigned to {j}")

Status: Optimal
Total Matching Score: 193.0

Assignments:
Carlos assigned to Tester
Joe assigned to Architect
Monika assigned to JavaDeveloper


------------------

### `networkx` Routine

In [11]:
# WRITE YOUR CODE HERE
# import libraries
import networkx as nx

Create the Graph

In [12]:
G = nx.DiGraph()

# add edges from Source to Resources
for r in resources:
    G.add_edge('Source', r, capacity=1, weight=0)

# add edges from Resources to Jobs
for (r, j), score in scores.items():
    G.add_edge(r, j, capacity=1, weight=-score)

# add edges from Jobs to Sink
for j in jobs:
    G.add_edge(j, 'Sink', capacity=1, weight=0)

Set Node Demands
- Source: Demand of -3 (Supply of 3 units)
- Sink: Demand of 3
- All other nodes: Demand of 0

In [13]:
demand = {'Source': -3, 'Sink': 3}
for node in G.nodes():
    if node not in demand:
        demand[node] = 0

nx.set_node_attributes(G, demand, 'demand')

Solve the MCNF Problem

In [14]:
# compute the minimum cost flow
flowCost, flowDict = nx.network_simplex(G)

Results

In [15]:
print("Total Matching Score:", -flowCost)

print("\nAssignments:")
for r in resources:
    for j in jobs:
        if flowDict[r][j] > 0.5:
            print(f"{r} assigned to {j}")

Total Matching Score: 193

Assignments:
Carlos assigned to Tester
Joe assigned to Architect
Monika assigned to JavaDeveloper


------------------
### Compare the two Solutions and Discuss Any Differences

**Observations from the Results**:  
From the output, we can observe that both the PuLP routine and the networkx routine yield the same total matching score of 193. Additionally, both methods assign resources to jobs in an identical manner.  
For this problem, both methods solve instantaneously and provide exact integer solutions due to the problem's structure and the integer capacities involved.

**Implementation Differences**  
The PuLP routine requires an explicit formulation of the linear programming problem, including defining variables, the objective function, and constraints. This method offers flexibility in modelling custom constraints and variables, making it highly adaptable to a wide range of problems. The solver used in PuLP depends on the default solver, such as CBC, but can be modified depending on the user’s preference.  
On the other hand, networkx utilises built-in network flow algorithms, such as network_simplex, which automatically handle flow conservation and capacity constraints based on the network structure. This makes it especially suitable for problems that can naturally be expressed as flow networks. Unlike PuLP, networkx requires less explicit formulation, relying instead on the inherent properties of network flow algorithms.

**Problem Relaxation**  
The constraint matrix in the assignment problem is
| Constraint / Variable | x₁,₁ | x₁,₂ | x₁,₃ | x₂,₁ | x₂,₂ | x₂,₃ | x₃,₁ | x₃,₂ | x₃,₃ | RHS |
|:----------------------|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:---:|
| Resource Constraints  |      |      |      |      |      |      |      |      |      |     |
| 1. Carlos             |  1   |  1   |  1   |  0   |  0   |  0   |  0   |  0   |  0   |  1  |
| 2. Joe                |  0   |  0   |  0   |  1   |  1   |  1   |  0   |  0   |  0   |  1  |
| 3. Monika             |  0   |  0   |  0   |  0   |  0   |  0   |  1   |  1   |  1   |  1  |
| Job Constraints       |      |      |      |      |      |      |      |      |      |     |
| 4. Tester             |  1   |  0   |  0   |  1   |  0   |  0   |  1   |  0   |  0   |  1  |
| 5. Java Developer     |  0   |  1   |  0   |  0   |  1   |  0   |  0   |  1   |  0   |  1  |
| 6. Architect          |  0   |  0   |  1   |  0   |  0   |  1   |  0   |  0   |  1   |  1  |

Where:

- x₁,₁: Carlos assigned to Tester.
- x₂,₂: Joe assigned to Java Developer.
- x₃,₃: Monika assigned to Architect.

which is totally unimodular.

In both implementations, variables are defined as binary to represent assignment decisions. The network simplex algorithm in networkx inherently works with integer flows when capacities and supplies/demands are integers, ensuring a natural alignment with the problem's requirements. This allows for direct interpretation of the solution in terms of assignments without needing additional rounding or post-processing steps.  
Relaxing the problem to allow fractional flows does not impact the optimality in this case due to the total unimodularity property of the assignment problem's constraint matrix. This property ensures that even when the problem is relaxed, the solution remains integer, guaranteeing that both the PuLP and networkx approaches deliver exact, optimal solutions.

**Sensitivity Analysis**  
PuLP allows access to dual variables and reduced costs if the solver provides them. Sensitivity analysis can be performed to understand how changes in matching scores affect the solution.  
In contrast, networkx does not directly provide dual variables. Sensitivity analysis would require re-running the algorithm with modified parameters.

**Certificates of Optimality**  
Both methods guarantee optimality for linear programming problems. PuLP provides detailed solver output, which can be used to analyze dual variables and perform sensitivity analysis.  
The networkx's network_simplex algorithm ensures that the flow cost is minimized, and given the problem's structure, the solution is optimal.